crates/net/discv5/src/lib.rs

```diff
@@ -0,0 +1,789 @@
+    ///
+    /// If the key already exists, this will update it.
+    ///
+    /// CAUTION: The value **must** be rlp encoded
+    pub fn set_eip868_in_local_enr(&self, key: Vec<u8>, rlp: Bytes) {
+        let Ok(key_str) = std::str::from_utf8(&key) else {
+            error!(target: "discv5",
+                err="key not utf-8",
+                "failed to update local enr"
+            );
+            return
+        };
+        if let Err(err) = self.enr_insert(key_str, &rlp) {
+            error!(target: "discv5",
+                %err,
+                "failed to update local enr"
+            );
+        }
+    }
+
+    /// Sets the pair in the EIP-868 [`Enr`] of the node.
+    ///
+    /// If the key already exists, this will update it.
+    pub fn encode_and_set_eip868_in_local_enr(
+        &self,
+        key: Vec<u8>,
+        value: impl alloy_rlp::Encodable,
+    ) {
+        let mut buf = BytesMut::new();
+        value.encode(&mut buf);
+        self.set_eip868_in_local_enr(key, buf.freeze())
+    }
+
+    /// Adds the peer and id to the ban list.
+    ///
+    /// This will prevent any future inclusion in the table
+    pub fn ban_peer_by_ip_and_node_id(&self, peer_id: PeerId, ip: IpAddr) {
+        match discv4_id_to_discv5_id(peer_id) {
+            Ok(node_id) => {
+                self.ban_node(&node_id, None);
+                self.ban_peer_by_ip(ip);
+            }
+            Err(err) => error!(target: "discv5",
+                %err,
+                "failed to ban peer"
+            ),
+        }
+    }
+
+    /// Adds the ip to the ban list.
+    ///
+    /// This will prevent any future inclusion in the table
+    pub fn ban_peer_by_ip(&self, ip: IpAddr) {
+        self.ban_ip(ip, None);
+    }
+
+    /// Returns the [`NodeRecord`] of the local node.
+    ///
+    /// This includes the currently tracked external IP address of the node.
+    pub fn node_record(&self) -> NodeRecord {
+        let enr: Enr<_> = EnrCombinedKeyWrapper(self.local_enr()).into();
+        (&enr).try_into().unwrap()
+    }
+
+    /// Spawns [`discv5::Discv5`]. Returns [`discv5::Discv5`] handle in reth compatible wrapper type
+    /// [`Discv5`], a receiver of [`discv5::Event`]s from the underlying node, and the local
+    /// [`Enr`](discv5::Enr) converted into the reth compatible [`NodeRecord`] type.
+    pub async fn start(
+        sk: &SecretKey,
+        discv5_config: Config,
+    ) -> Result<(Self, mpsc::Receiver<discv5::Event>, NodeRecord), Error> {
+        //
+        // 1. make local enr from listen config
+        //
+        let Config {
+            discv5_config,
+            bootstrap_nodes,
+            fork,
+            tcp_port,
+            other_enr_data,
+            lookup_interval,
+            discovered_peer_filter,
+        } = discv5_config;
+
+        let (enr, bc_enr, ip_mode, fork_id_key) = {
+            let mut builder = discv5::enr::Enr::builder();
+
+            use discv5::ListenConfig::*;
```

Comment:
let's move this to the top of the file, I prefer to keep imports out of functions